proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+}
+```
+
+
+### Implementing Scripting Support
+     
+To implement (custom) scripting support one should do the following
+
+1. Implement common (IDE and host) parts
+   - Choose target platform or implement custom one
+   - choose appropriate preprocessor or implement custom
+   - choose/implement selector
+   - define script class with the `@ScriptDefinition` annotation
+   - implement executor for the script class
+
+2. Implement the scripting host with the following functionality:
+   - load script definition class
+   - initialize the platform (e.g. setup java home, initial dependencies, etc)
+   - initialize preprocessor with the platform (e.g. for using compilation services)
+   - initialize executor
+   - initialize selector
+   - for each script:
+     - (optionally) - check if the script is supported
+     - call preprocessor
+     - call platform to compile script
+     - call execute to get an instance
+     - call executor to run the script instance
+   - *(the base host may implement most of the boilerplate)*
+      
+3. Implement IDE support host with the following functionality
+   - load script definition class
+   - initialize the platform (e.g. setup java home, initial dependencies, etc)
+   - initialize preprocessor with the platform (e.g. for using compilation services)
+   - initialize selector
+   - provide these components to the IDE
+
+### IDE Interface
+
+In the IDE the support for rich editing could be implemented using the specific extension point:
+
+```
+interface CustomKotlinScriptingProvider {
+    val scriptDefinition: kClass<*>
+    val platform: ScriptingPlatform<CompilationConfig>
+    val preprocessor: ScriptPreprocessor<CompilationConfig>
+    val selector: ScriptSelector
+}
+```
+
+### Standard hosts and discovery
+
+For standard JVM compiler, a command line parameter could be used to specify a script definition class name, and regular 
+compilation classpath could be used for dependencies.
+
+Additionally, the automatic discovery of the templates marked by `@ScriptDefinition` annotation could be used with 
+libraries that do not have plugins able to provide an extension. This could, for example, be used for test frameworks.
+
+To find the script definition corresponding to the script file, the compiler uses one of the following methods:
+
+* Use the script definition specified explicitly in CompilerConfiguration (the Gradle case);
+* Scan for all .jar in the classpath, load the list of script template definitions in each .jar from the JAR metadata, 
+  and detect the applicable one based on the `@ScriptFilePattern` annotation;
+* Use an explicit annotation referring to the FQ name of the script definition class:
+
+```
+@file:ScriptTemplate("org.jetbrains.kotlin.gradle.GradleScript")
+```
+
+### Compilation Configuration
+
+The `CompilationConfig` for the JVM platform may look like:
+```
+interface KotlinJVMScriptAdditionalParameters {
+    val classpath: Iterable<File> // dependencies classpath
```

Comment:
Yes, anything that platform needs, this is just an example. The actual definition of the `CompilationConfig` should not be specified, but rather taken from the platform definition.